packages/require/lib/types.ts

```diff
@@ -0,0 +1,112 @@
+import TruffleConfig from "@truffle/config";
+import Resolver from "@truffle/resolver";
+import TruffleError from "@truffle/error";
+
+export interface RequireOptions {
+  /**
+   * The path to the source file to be required/executed
+   */
+  file: string;
+
+  /**
+   * The `@truffle/config` for the user's project
+   */
+  config: TruffleConfig;
+}
+
+export interface ExecOptions {
+  contracts_build_directory: string;
+  /**
+   * Path to file to execute. Must be a module that exports a function.
+   */
+  file: string;
+
+  /**
+   * Object containing any global variables you'd like set when this function is
+   * run.
+   */
+  context: {
+    [index: string]: any;
+  };
+
+  /**
+   * The instance of `@truffle/resolver` to be passed into the global context
+   * when executing the script
+   */
+  resolver: Resolver;
+
+  /**
+   * The JSON RPC provider that will be used by the interface-adpater and web3
+   * instance injected into the script
+   */
+  provider: Provider;
+
+  /**
+   * A reference to a network defined in the user's `@truffle/config`.
+   */
+  network: string | number;
+
+  /**
+   * The id of the network as reported by the `net_version` JSON RPC method.
+   */
+  network_id: string | number;
+
+  /**
+   * The configuration for any networks that may be referenced by the script
+   * being executed.
+   */
+  networks: {
+    [index: string]: {
+      type: string; // likely "ethereum", but not guaranteed
+    };
+  };
+}
+
+/**
+ * The body of a valid JSON RPC request
+ */
+interface JsonRpcRequest {
+  jsonrpc: string;
+  method: string;
+  params: any[];
+  id: number;
+}
+
+/**
+ * The body of a valid JSON RPC response
+ */
+export interface JsonRpcResponse {
+  jsonrpc: string;
+  id: number;
+  result?: any;
+  error?: string;
```

Comment:
I'd agree, however this needs to be permissive to providers from Web3, and web3 doesn't have this restriction.

Reply:
Thanks again for the quick references. I really wish the web3 codebase would finish converting over to TypeScript so these types were all generated from source. The amount of change from version-to-version is a bit annoying.

Reply:
Figured out where the weirdness came from. We're depending on `@types/web3@1.0.20` in a number of our packages. Of course the types in that package are totally wrong, and converting to the correct types here causes compilation errors. 😞

I'll figure something out, though.

Reply:
OK - got that sorted. Wound up using the `provider` type from `web3-core/types` (but imported/exported it as `Provider` because who names their types with a lower case first letter?)